File: src/onion/compiler/Symbol.java
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2005, Kota Mizushima, All rights reserved.       *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler;

import java.util.HashMap;
import java.util.Map;

/**
 * 
 * @author Kota Mizushima
 *
 */
public final class Symbol {
  private static Map symbols = new HashMap();
  
  public static Symbol intern(String name) {
    Symbol symbol = (Symbol)symbols.get(name);
    if(symbol == null){
      symbol = new Symbol(name);
      symbols.put(name, symbol);
    }
    return symbol;
  }
  
  private String name;
  
  /**
   * Users cannot create Symbol object directly.
   * @param name
   */
  private Symbol(String name) {
    this.name = name;
  }

  /**
   * This method returns this object's name.
   * @return this object's name
   */
  public String getName() {
    return name;
  }
}

File: src/main/scala/onion/compiler/Symbol.scala
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2016-, Kota Mizushima, All rights reserved.  *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler

import collection.mutable.HashMap

/**
 *
 * @author Kota Mizushima
 *
 */
object Symbol {
  private val symbols = new HashMap[String, Symbol]
  def apply(name: String): Symbol = {
    symbols.get(name) match {
      case Some(symbol) => symbol
      case None =>
        val symbol = new Symbol(name)
        symbols(name) = symbol
        symbol
    }
  }

}

/**
 * Users cannot create Symbol object directly.
 * @param name
 */
final class Symbol private (val name: String) extends AnyRef {
  override def equals(obj: Any): Boolean = this eq obj.asInstanceOf[AnyRef]
  override def hashCode(): Int = super.hashCode()
}